aloha-core/src/main/scala/com/eharmony/aloha/models/multilabel/MultilabelModel.scala

```diff
@@ -0,0 +1,495 @@
+  */
+case class MultilabelModel[U, K, -A, +B <: U](
+    modelId: ModelIdentity,
+    featureNames: sci.IndexedSeq[String],
+    featureFunctions: sci.IndexedSeq[GenAggFunc[A, Sparse]],
+    labelsInTrainingSet: sci.IndexedSeq[K],
+    labelsOfInterest: Option[GenAggFunc[A, sci.IndexedSeq[K]]],
+    predictorProducer: SparsePredictorProducer[K],
+    numMissingThreshold: Option[Int],
+    auditor: Auditor[U, Map[K, Double], B])
+(implicit ev: SerializabilityEvidence[K])
+extends SubmodelBase[U, Map[K, Double], A, B]
+   with RegressionFeatures[A] {
+
+  import MultilabelModel._
+
+  /**
+    * predictory is transient lazy value because we don't need to worry about serialization.
+    * We don't care about the lazy property.  It should be created eagerly.
+    */
+  @transient private[this] lazy val predictor = predictorProducer()
+  predictor // Force predictor eagerly
+
+  /**
+    * Cache this in case labelsOfInterest is None.  In that case, we don't want to repeatedly
+    * create this because it could create a GC burden for no real reason.
+    */
+  private[this] val defaultLabelInfo =
+    // Hopefully, making this non-transient doesn't increase the serialized size much.
+    // labelsInTrainingSet isn't serialized twice is it?
+    LabelsAndInfo(labelsInTrainingSet.indices, labelsInTrainingSet, Seq.empty, None)
+
+  /**
+    * Making from label to index into the sequence of all labels encountered during training.
+    */
+  private[this] val labelToInd: Map[K, Int] =
+    labelsInTrainingSet.zipWithIndex.map { case (label, i) => label -> i }(collection.breakOut)
+
+  override def subvalue(a: A): Subvalue[B, Map[K, Double]] = {
+    val li = labelsAndInfo(a, labelsOfInterest, labelToInd, defaultLabelInfo)
+
+    if (li.labels.isEmpty)
+      reportNoPrediction(modelId, li, auditor)
+    else {
+      val Features(x, missing, missingOk) = constructFeatures(a)
+
+      if (!missingOk)
+        reportTooManyMissing(modelId, li, missing, auditor)
+      else {
+        // TODO: To support label-dependent features, fill last parameter with a valid value.
+        val predictionTry = predictor(x, li.labels, li.indices, sci.IndexedSeq.empty)
+
+        predictionTry match {
+          case Success(pred) => reportSuccess(modelId, li, missing, pred, auditor)
+          case Failure(ex)   => reportPredictorError(modelId, li, missing, ex, auditor)
+        }
+      }
+    }
+  }
+
+  /**
+    * When the `predictor` passed to the constructor is a java.io.Closeable, its `close`
+    * method is called.
+    */
+  override def close(): Unit =
+    predictor match {
+      case closeable: Closeable => closeable.close()
+      case _ =>
+    }
+}
+
+object MultilabelModel extends ParserProviderCompanion {
+
+  /**
+    * Contains information about the labels to be used for predictions, and problems encountered
+    * while trying to get those labels.
+    * @param indices indices into the sequence of all labels seen during training.  These should
```

Comment:
If these are parallel arrays, then why not use an `IndexedSeq[(Int, K)]` to force the same length in the type?